pkg/controller/core/workload_controller.go

```diff
@@ -144,11 +154,29 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	log.V(2).Info("Reconciling Workload")
 
 	if len(wl.ObjectMeta.OwnerReferences) == 0 && !wl.DeletionTimestamp.IsZero() {
+		// manual deletion triggered by the user
 		return ctrl.Result{}, workload.RemoveFinalizer(ctx, r.client, &wl)
 	}
 
-	if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadFinished) {
-		return ctrl.Result{}, nil
+	finishedCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadFinished)
+	if finishedCond != nil && finishedCond.Status == metav1.ConditionTrue {
+		if r.objectRetention == nil  {
+			return ctrl.Result{}, nil
+		}
+		now := r.clock.Now()
+		expirationTime := finishedCond.LastTransitionTime.Add(r.objectRetention.Duration)
+		if now.After(expirationTime) {
+			log.V(2).Info("Deleting workload because it has finished and the retention period has elapsed", "retention", r.objectRetention.Duration)
+			if err := r.client.Delete(ctx, &wl); err != nil {
+				log.Error(err, "Failed to delete workload from the API server")
+				return ctrl.Result{}, fmt.Errorf("deleting workflow from the API server: %w", err)
+			}
+			r.recorder.Eventf(&wl, corev1.EventTypeNormal, "Deleted", "Deleted finished workload due to elapsed retention:  %v", workload.Key(&wl))
```

Comment:
```suggestion
			r.recorder.Eventf(&wl, corev1.EventTypeNormal, "Deleted", "Deleted finished workload due to elapsed retention")
```
Because the event is already targeting the workload, we don't need to add it's name again in the message.

Reply:
Removed